File: 3. Longest Substring Without Repeating Characters/solution1/Solution.java
import java.util.HashMap;

class Solution {
  public int lengthOfLongestSubstring(String s) {
    HashMap<Character, Integer> charMap = new HashMap();
    int l = 0, maxLen = 1;
    for (int r = 0; r < s.length(); r++) {
      if (charMap.containsKey(s.charAt(r))) {
        maxLen = Math.max(maxLen, r - l);
        int nextL = charMap.get(s.charAt(r)) + 1;
        while (l < nextL) {
            charMap.remove(s.charAt(l));
            l++;
        }
      }
      charMap.put(s.charAt(r), r);
    }
    // Always check the last substring till the end.
    return s.length() == 0 ? 0 : Math.max(maxLen, s.length() - l);
  }
}

File: 3. Longest Substring Without Repeating Characters/solution1/solution.cc
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    unordered_map<char, int> charMap;
    int l = 0, maxLen = 0;
    for (int r = 0; r < s.size(); ++r) {
      if (charMap.count(s[r])) {
        maxLen = max(maxLen, r - l);
        int nextL = charMap[s[r]] + 1;
        while (l < nextL) {
          charMap.erase(s[l]);
          ++l;
        }
      }
      charMap[s[r]] = r;
    }
    return s.empty() ? 0 : max(maxLen, (int)s.size() - l);
  }
};